### DISCUSSION TITLE
Testing pipeline with scrapy shell

### DISCUSSION THREAD

**Discussion by alberand** [scrapy/scrapy#6592] [Q&A]
Hi all,

First of all, thanks for amazing project!

With scrapy shell I can run `scrapy shell <url>` to play around with the way spider parses the page. Is there a way to run full scrapy processing on a single `<url>`?

My pipeline transforms data in the `item`, does a few database updates/quieries. Testing this is quite painful - running the whole spider which start processing all the available pages. I would like to run pipeline in the `scrapy shell` for a single item, is there a way? I know I can manipulate with start urls and limit spider to stop but this is plan B.

**Comment by Gallaecio** [comment]
I’m pretty sure there’s no built-in way to run pipelines like that. I’m turning this into a feature request.

Could you clarify what you would like, exactly? Would you want to hard-code an item and run the pipelines for that item? Would you prefer to choose a URL and callback as inputs and get pipelines to run on the output of that?

(For the output file, I guess an `-o/-O` parameter like in `scrapy crawl` would make sense)

**Reply by alberand** [reply]
For shell, maybe something like this with new pipelines object:
```
>>> item = pipelines.process(spider.parse(response))
...
run all pipelines in the ITEM_PIPELINES
...
>>> item = pipelines['scrapy.pipelines.images.ImagesPipeline'].process(spider.parse(response))
...
run only `scrapy.pipelines.images.ImagesPipeline` pipeline
...
```
and for `cli` arguments `-p/--pipeline [PIPELINE_NAME]` could be used, with PIPELINE_NAME being empty means run all.

I not fully familiar with scrapy architecture, so not sure which interface is more suitable.   

> Would you want to hard-code an item and run the pipelines for that item? 

Will work if it's possible to just pass `spider.parse(response)` item

> Would you prefer to choose a URL and callback as inputs and get pipelines to run on the output of that?

Basically the same and will also work.

> Could you clarify what you would like, exactly?

My goal is to be able to test pipeline inputs and outputs. Same as we can do with a spider, we can look into request and response then run the spider and check that everything works as expected. This is especially useful when spider fails to parse a particular page/url, due to wrong xpath, for example.

With pipelines, it much more difficult to debug something like this. Spider can successfully parse the page and generate an item. Then, for example, regex in the pipeline incorrectly processes one of the fields. At the end of the pipeline when we want to store the item, it has incorrect data. Rerunning pipeline on the same url/item would be great to debug the problem.

**Comment by GeorgeA92** [comment]
@Gallaecio 

> I’m pretty sure there’s no built-in way to run pipelines like that. I’m turning this into a feature request.

As mentioned on https://github.com/scrapy/scrapy/issues/6554#issuecomment-2517779776 it is possible to update/reimplement logic of shell command to add custom processing logic for each incoming response on current versions of scrapy it can be calling pipeline method, calling item_proc (all `process_item` methods chain from pipelines ) or anything else/

This option exists from 2009 as I see from source code, may be it is not documented well. As this option is possible to implement by changing settings and set custom class from existing scrapy components (almost as setting custom middlewares/pipelines/extensions etc.) on current and very old versions of scrapy - NO additional updates to scrapy source required to implement this. 

@alberand 

> With scrapy shell I can run `scrapy shell <url>` to play around with the way spider parses the page. Is there a way to run full scrapy processing on a single `<url>`?

Running spider on single url to test pipelines and other project related things - is quite typical request for projects.
Typical industry practice for this cases is to update spider class to make possible call something like `scrapy crawl spider_name -a product_url=<url>` where single url received from cli argument processed that produce items(from existing spider's callback method) that later processed by spidermiddlewares and pipelines / feedexporters as on regular crawl (which indeed is not happening on `scrapy shell` command that didn't aimed for this)

**Comment by alberand** [comment]
For anyone else looking for a reference implementation of solution suggested above, this should be enough to be able to run spider on a single URL:
```python
def start_requests(self):
    if self.product_url:
        return [ scrapy.FormRequest(self.product_url, callback=self.parse_item) ]

    return [ scrapy.FormRequest(url, callback=self.parse)  for url in self.start_urls]
```
You need to set `self.item_url` somehow, though. I do it in `self.from_crawler()` and then run spiders with `CrawlerProcess(get_project_settings()).crawl(spider, item_url=argparse.product_url)`. You can use spider argument if you run them directly https://docs.scrapy.org/en/latest/topics/spiders.html#spiderargs

```python
class BaseSpider(scrapy.Spider):
    @classmethod
    def from_crawler(cls, crawler, *args, **kwargs):
        spider = super(BaseSpider, cls).from_crawler(crawler, *args, **kwargs)

        # We want to parse a single URL
        if 'product_url' in kwargs.keys():
            spider.product_url = kwargs['product_url']
        else:
            spider.product_url = None

        return spider

```
This has nothing to do with `scrapy shell`, though.
